TextToSpeech/impl/TTSSpeaker.cpp

```diff
@@ -885,53 +984,42 @@ std::string TTSSpeaker::constructURL(TTSConfiguration &config, SpeechData &d) {
         return "";
     }
 
-    // EndPoint URL
-    std::string tts_request;
-    TTSURLConstructer url;
-    tts_request = url.constructURL(m_defaultConfig,d.text,false);
-    TTSLOG_WARNING("Constructured final URL is %s", tts_request.c_str());
+    TTSURLConstructer urlConstructor;
+    std::string tts_request = urlConstructor.constructURL(m_defaultConfig, d.text, false, shouldUseLocalEndpoint());
+    if(m_defaultConfig.hasValidLocalEndpoint()) {
+       PipelineType pipelineType = getUrlPipelineType(tts_request);
+       if(pipelineType != m_pipelinetype) {
+          //pipeline switch required
+          TTSLOG_INFO("pipeline needs updation");
+          destroyPipeline();
+          createPipeline(pipelineType);
+       } else {
+          TTSLOG_INFO("re-use existing pipeline.");
+       }
+    }
+    TTSLOG_WARNING("Constructed final URL is %s", tts_request.c_str());
     return tts_request;
 }
 
 void TTSSpeaker::speakText(TTSConfiguration config, SpeechData &data) {
     m_isEOS = false;
     m_duration = 0;
 
-    if(m_pipeline && !m_pipelineError && !m_flushed) {
+    if((m_pipeline && !m_pipelineError && !m_flushed) || m_remoteError) {
+        string url;
```

Comment:
I would consider moving m_currentSpeech update to one place play()

Reply:
```
void TTSSpeaker::speakText(TTSConfiguration config, SpeechData &data) {
    m_isEOS = false;
    m_duration = 0;

    if(m_pipeline && !m_flushed) {
        string token;
        bool authrequired = (config.endPointType().compare("TTS2") == 0);
        if(authrequired)
            token = WPEFramework::Plugin::TTS::SatToken::getInstance(config.satPluginCallsign())->getSAT();

        play(constructURL(config, data), data, authrequired, token);
    } else {
        TTSLOG_WARNING("m_pipeline=%p, m_flushed=%d, m_pipelineError=%d", m_pipeline, m_flushed, m_pipelineError);
    }
}

void TTSSpeaker::play(string url, SpeechData &data, bool authrequired, string token) {
    m_currentSpeech = &data;

    g_object_set(G_OBJECT(m_source), "location", url.c_str(), NULL);

    if(authrequired) {
        string authStr = "test, Authorization=(string)\"Bearer\\ " + token + "\"";
        GstStructure* extraHeaders = gst_structure_from_string((const gchar*)(authStr.c_str()), NULL);
        if (extraHeaders != NULL) {
            g_object_set(G_OBJECT(m_source), "extra-headers", extraHeaders, NULL);
            gst_structure_free(extraHeaders);
        }
    }

    // PCM Sink seems to be accepting volume change before PLAYING state
    g_object_set(G_OBJECT(m_audioVolume), "volume", (double) (data.client->configuration()->volume() / MAX_VOLUME), NULL);

    gst_element_set_state(m_pipeline, GST_STATE_PLAYING);

    #if defined(PLATFORM_AMLOGIC) || defined(PLATFORM_REALTEK)
    setMixGain(MIXGAIN_PRIM,data.primVolDuck);
    #endif
    TTSLOG_VERBOSE("Speaking.... ( %d, \"%s\")", data.id, data.text.c_str());

    //Wait for EOS with a timeout incase EOS never comes
    if(m_pcmAudioEnabled) {
        //FIXME, find out way to EOS or position for raw PCM audio
        waitForAudioToFinishTimeout(60);
    }
    else {
        waitForAudioToFinishTimeout(10);
    }

    m_currentSpeech = NULL;
}
```

Reply:
would be nice to keep related functions close for e.g speakText() & play() and the other dependencies